### ISSUE TITLE
:col value off-by-one for some vars in :var-usages portion of analysis data

### ISSUE THREAD

**Issue by sogaiu** [borkdude/clj-kondo#568]
**version**

commit 9de6b455a8ccd6df7923cb480b621c42572908a1

**platform**

manjaro linux

**problem**

At least some used identifiers are reported as being at a column which is one less than they should be.

For example, an identifier might be located at column 2, but is reported as being at column 1.

**repro**

`clj-kondo --lint - --config '{:output {:analysis true :format :edn}}}' <<< '(+ 1)' | jet --pretty --query ':analysis :var-usages'`

```
[{:fixed-arities #{0 1 2},
  :name +,
  :filename "<stdin>",
  :from user,
  :col 1,
  :arity 1,
  :varargs-min-arity 2,
  :row 1,
  :to clojure.core}]
```

**expected behavior**

In the above reproduction, I expected to see column 2 for + because it occurs immediately after the first character in the string.

(According to the main developer, internally clj-kondo uses 1-based row and column positions.)

**actual behavior**

I observed a column value of 1 for +.

**notes**

Below are some notes on some related functions and data:

The entry point `core/run!` eventually leads to two areas of interest:

* `reg-var-usage!` - stores usage info while linting -- into ctx's :used-vars
* `reg-usage!` - stores usage info for the analysis output -- from ctx's :used-vars

Perhaps one might say the two functions communicate or are linked via ctx's :used-vars.  Note that there may be multiple :used-vars within ctx, but perhaps without loss of generality, one can reason as if there is only one for the current situation.

`reg-var-usage!` is reached by:

* `core/run!`
* `impl.core/process-files`
...
* `impl.analyzer/analyze-expression**`
* `impl.analyzer/analyze-call`
* `impl.namespace/reg-var-usage!`

`reg-usage!` is reached by:

* `core/run!`
* `impl.linters/lint-var-usage`
* `impl.analysis/reg-usage!`

What gets stored by `reg-var-usage!` is provided by `analyze-call` and is roughly:

```
                call (cond-> {:type :call
                              :resolved-ns resolved-namespace
                              :ns ns-name
                              :name (with-meta
                                      (or resolved-name full-fn-name)
                                      (meta full-fn-name))
                              :unresolved? unresolved?
                              :clojure-excluded? clojure-excluded?
                              :arity arg-count
                              :row row
                              :col (inc col)
                              :base-lang base-lang
                              :lang lang
                              :filename (:filename ctx)
                              :expr expr
                              :callstack (:callstack ctx)
                              :config (:config ctx)
                              :top-ns (:top-ns ctx)
                              :arg-types (:arg-types ctx)}
                       id (assoc :id id)
                       in-def (assoc :in-def in-def))]
```

So :used-vars contains information about calls -- including their locations (:row and :col).

`reg-usage!` is basically handed the :row and :col values obtained by `impl.linters/lint-var-usage` from :used-vars (i.e. including :row and :col values that correspond to calls).

That doesn't seem quite right, though it should be quite close.  The :col value may be off-by-one in many (or all?) cases -- the call :col value is (typically?) for a left paren, while the desired :col value is for the symbol in the so-called function-call position (one to the right of the left paren).  For example:

```
(+ 1 1)
```

The left-most paren corresponds to the call's :col value.

The + corresponds to the symbol's :col value.

For calls, perhaps instead of:

```
(analysis/reg-usage! ctx
                     filename row col caller-ns-sym
```

we could do instead:

```
(analysis/reg-usage! ctx
                     filename row
                     (if call? (and col (inc col))
                       col)
                     caller-ns-sym
```

That is, distinguish between calls and non-calls.  For calls, increment the col value (if it's defined), but when it's not a call, col should be left alone.

I'm not confident yet this will always work, but it's not completely untested either :)  I don't know how the following sort of call is handled by clj-kondo:

```
((fun-generator) 1 2)
```

That is, for the outer call, there is no symbol in the so-called function-call position.  May be this is a non-issue?

How does the above description seem?

**Comment by borkdude** [comment]
```
   (if call? (and col (inc col))
                       col)
```

This is bad practice, since the call may look like this: `(   + 1 2 3)`.
Instead we should output the row and col of the name used in the call to the analysis export while leaving the rest of the code untouched. You can get to the row and col of the name via metadata on the name symbol.
